[net] Cleanup connection types- followups

amitiuttarwar opened the pull request:
This PR addresses outstanding review comments from #19316. It further simplifies `net.cpp` complexity and adds documentation about design goals about different connection types.

amitiuttarwar committed:
[net] Remove unnecessary default args on OpenNetworkConnection

amitiuttarwar committed:
[refactor] Restructure logic to check for addr relay.

We previously identified if we relay addresses to the connection by checking
for the existence of the m_addr_known data structure. With this commit, we
answer this question based on the connection type.

IsAddrRelayPeer() checked for the existence of the m_addr_known

amitiuttarwar committed:
[trivial] Small style updates

amitiuttarwar committed:
[test] Add explicit tests that connection types get set correctly

amitiuttarwar committed:
scripted-diff: Rename `OUTBOUND` ConnectionType to `OUTBOUND_FULL_RELAY`

-BEGIN VERIFY SCRIPT-
sed -i 's/OUTBOUND, /OUTBOUND_FULL_RELAY, /g' src/net.h
sed -i 's/ConnectionType::OUTBOUND/ConnectionType::OUTBOUND_FULL_RELAY/g' src/test/net_tests.cpp src/test/fuzz/process_message.cpp src/test/fuzz/process_messages.cpp src/net.cpp src/test/denialofservice_tests.cpp src/net.h src/test/fuzz/net.cpp
-END VERIFY SCRIPT-

amitiuttarwar committed:
[refactor] Simplify check for block-relay-only connection.

Previously we deduced it was a block-relay-only based on presence of the
m_tx_relay structure. Now we have the ability to identify it directly via a
connection type accessor function.

amitiuttarwar committed:
[refactor] Simplify connection type logic in ThreadOpenConnections

Consolidate the logic to determine connection type into one conditional to
clarify how they are chosen.

amitiuttarwar committed:
[doc] Describe connection types in more depth.

amitiuttarwar commented:
I'm not 100% convinced about replacing `IsAddrRelayPeer()` with `RelayAddrsWithConn()`  9345f9c9d47a88987606d1421823914a4c48dbf9. 

Pros: 
- It is more direct to check the connection type instead of a (missing) data structure based on the connection type.
- Less mental overhead, code touch points to glean intention. 

Cons: 
- We are now checking the same conditional in two places- [[1](https://github.com/bitcoin/bitcoin/pull/19724/commits/9345f9c9d47a88987606d1421823914a4c48dbf9#diff-1a8b9d1ad0a6fda5e751286c73102fc2R867)] & [[2](https://github.com/bitcoin/bitcoin/blob/master/src/net.cpp#L2783)], and need to manually ensure they do not fall out of sync.

sipa commented:
@amitiuttarwar I think it makes sense to make the addr-relay decision depend on the connection type. That should be the most authoritative information about the connection. Sure, it's duplication, but if we accidentally forget to create the necessary data structures, test will fail; if we accidentally create them if they're not needed, at worst we've wasted a bit of memory we're already ok with for other nodes.

MarcoFalke commented on src/net.cpp:
```diff
@@ -1841,11 +1841,11 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
                 // also have the added issue that they could be attacker controlled and used
                 // to prevent us from connecting to particular hosts if we used them here.
-                switch(pnode->m_conn_type){
+                switch (pnode->m_conn_type){
```
style-nit:

```suggestion
                switch (pnode->m_conn_type) {
```

if you decide to do style-fixups, it would probably be good to do all of them in one go or not at all. Maybe with clang-format-diff?   https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy

amitiuttarwar replied:
heh. thanks for the nudge. I've updated my tooling so I'll pay better attention to the clang-format-diff moving forward. I think I've now cleaned everything up properly in this PR? I'm not sure what exactly you mean by "all in one go or not at all"... 

(disclaimer: I didn't take the suggestions to not [IndentCaseBlocks](https://clang.llvm.org/docs/ClangFormatStyleOptions.html). I would much prefer if we turned that format option to true.)

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2433,9 +2433,8 @@ void ProcessMessage(
         }
 
-        if (!pfrom.IsInboundConn() && pfrom.IsAddrRelayPeer())
+        if (pfrom.AdvertiseAddressToConn())
```
If we're cleaning this up, we should really clean it up and move the `connman.MarkAddressGood()` out of this if statement, since it's not related to address advertising at all. Having it in the same conditional is a trap for developers to fall into.

(specifically, a developer might think "we don't need to advertise our address to FEELER connections, since we disconnect before we actually send the ADDR or GETADDR messages, so we can remove FEELER from AdvertiseAddressToConn", but doing that breaks our FEELER connection logic, because `MarkAddressGood()` needs to be called to mark the address as good)

cc @sdaftuar

sdaftuar replied:
I was looking at this more and I believe we should leave the MarkAddressGood call in the if-block, though it needs an explanation— the reason (I think) is that we don’t want block-relay connections to leave a trace in the addrman that could be leaked by an attacker sniffing our addr responses.

So this just needs a comment explaining why it’s in here.

jnewbery replied:
Can you explain this a bit? I think you're saying that if we connect out to a peer that we use for block-relay-only we don't want to mark it as good (essentially move it from New to Tried and update the last successful connect time) because then another peer might be able to tell that we've connected to them?

My understanding of how we respond to `getaddr` messages is that we'll respond with (up to) 1000 records taken at random from all records, regardless of whether they're in New or Tried, so marking a peer as good won't change whether we include it in `addr` responses.

sdaftuar replied:
I don't have a great understanding of the details of how addrman works or how we interact with it -- so possibly the fear was unfounded.  It used to be the case that you could trivially download a listening node's addrman though, and it also used to be the case (perhaps only long ago?) that we'd leak information about current connections based on time stamps stored in our addrman, so when I worked on #15759 I tried to [leave the addrman state unchanged](https://github.com/bitcoin/bitcoin/pull/15759#issuecomment-480398802) by making block-relay-only connections.

I'm not sure that I fully succeeded in not leaking information, or that the concern is correct to begin with given how addrman already works.  So it's possible that with some analysis we could make improvements here.  Maybe not appropriate for this pr though...

jnewbery replied:
I believe there are three ways that a peer's record is updated in addrman when we're connected to them:

1. when we receive a `version` message from them, we'll mark their address good. That moves the address from the new table to the tried table, and updates the `nLastSuccess`/`nLastTry`/`nAttempts` fields. We only do this if the peer isn't inbound and isn't block-relay-only. I don't believe this change can easily be detected by a peer sending us `getaddr` messages.
2. if a peer self-advertises to us, we'll call `AddAddressKnown()`, which could update the `nTime`, which is included in the address record that we send to a peer in response to a `getaddr` request.
3. when we disconnect from the peer, we'll update the `nTime`, as long as the peer was marked as `fCurrentlyConnected`. That happens when we receive a `verack` from a peer which is not inbound. This doesn't include feeler peers, since we disconnect from them as soon as we receive a `version` message from them. The `nTime` is only updated during disconnection so spies can't find out who we're currently connected to (https://github.com/bitcoin/bitcoin/pull/5860#issuecomment-78397140).

In summary, I think block-relay-only peers _should_ have their address marked good when we connect to them (1). If we don't, then we'll update the `nLastTry` and `nAttempts` fields when we attempt to connect to them, but not update the `nLastSuccess`.

The reason it's important in this PR is that placing `MarkAddressGood()` behind a conditional `if (AdvertiseAddressToConn())` is very confusing for anyone reading the code, and makes it easy to introduce subtle bugs.

amitiuttarwar replied:
I agree that having `MarkAddressGood` in this conditional is surprising & that changing to `AdvertiseAddressToConn` augments the confusion. 

For this PR- I've updated the conditional to be `!pfrom.IsInboundConn() && !pfrom.IsBlockOnlyConn()` (aka removed the misleading function), and added a comment.

For future work- I've noted this down. I've taken an initial look and @jnewbery I believe what you're saying makes sense, but I'll have to spend more time with it to fully convince myself. The fact that I missed this behavior in my initial pass affirms that its easy to miss or confuse. I'd like to either (1) understand why it should be in this conditional & comment or (2) understand why it shouldn't be & change it. But I'm leaving this for another PR :)

naumenkogs replied:
I think we're taking the wrong approach in the PR as it is right now. *At least* that condition should be split in 2 blocks:

I think the current code change here is not useful, but only distracts:
I) The comment `Advertise our address` is removed for no reason
II) The comment `(unrelated to advertising our address)` is very confusing. I managed to understand it only after reading this discussion
III) The comment `Moves address from New to Tried table in Addrman` is [not sufficient](https://github.com/bitcoin/bitcoin/pull/19724/files/c343ba35cfecb8aaf3aecb1259b6b30a225507f3..f26b0ddac590ac96e86fe043de744747e834ec0e#r480969871)
IV) The change of `IsAddrRelayPeer` to `IsBlockOnlyConn` [here](https://github.com/bitcoin/bitcoin/pull/19724/files/c343ba35cfecb8aaf3aecb1259b6b30a225507f3..f26b0ddac590ac96e86fe043de744747e834ec0e#r480966499) is very confusing, the two blocks should be split into 2 very separate blocks according to the following reasoning:

1. We self-announce to every peer that supports self-announcement AND it's a good idea:
1a) not useful to self-announce to feelers as we won't actually send ADDR to them (i think?)
1b) bad to self-announce to block-relay-only (maybe? wrt privacy)
2. We MarkAddressGood according to the logic we have

naumenkogs replied:
w.r.t block-relay-only peers and AddrMan, I agree it's a separate issue, so i'd prefer to preserve the approach status quo here.

The current PR currently satisfies it, but I'm still unhappy with the code changes in this particular area (see the previous comment).

sdaftuar replied:
@naumenkogs Would it be enough to just add some more comments to this block?  I think it's fine to leave as one block until we actually need to split it, so I'd propose something like this:

```diff
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 7cd4f7c4cb5..4285e8ba898 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -2427,6 +2427,21 @@ void PeerLogicValidation::ProcessMessage(CNode& pfrom, const std::string& msg_ty
         }
 
         if (!pfrom.IsInboundConn() && !pfrom.IsBlockOnlyConn()) {
+            // For outbound peers, we try to relay our address (so that other
+            // nodes can try to find us more quickly, as we have no guarantee
+            // that an outbound peer is even aware of how to reach us) and do a
+            // one-time address fetch (to help populate/update our addrman). If
+            // we're starting up for the first time, our addrman may be pretty
+            // empty and no one will know who we are.
+            //
+            // We also will update the addrman to record connection success for
+            // these peers (which include OUTBOUND_FULL_RELAY peers and FEELER
+            // connections) so that addrman will have an up-to-date notion of
+            // which peers are online and available.
+            //
+            // We skip these operations for BLOCK_RELAY peers to avoid
+            // potentially leaking information about our BLOCK_RELAY
+            // connections via the addrman or address relay.
             if (fListen && !::ChainstateActive().IsInitialBlockDownload())
             {
                 CAddress addr = GetLocalAddress(&pfrom.addr, pfrom.GetLocalServices());
@@ -2446,8 +2461,8 @@ void PeerLogicValidation::ProcessMessage(CNode& pfrom, const std::string& msg_ty
             m_connman.PushMessage(&pfrom, CNetMsgMaker(nSendVersion).Make(NetMsgType::GETADDR));
             pfrom.fGetAddr = true;
 
-            // Moves address from New to Tried table in Addrman
-            // (unrelated to advertising our address)
+            // Moves address from New to Tried table in Addrman, resolves
+            // tried-table collisions, etc.
             m_connman.MarkAddressGood(pfrom.addr);
         }
```

naumenkogs replied:
I'd suggest making `If we're starting up for the first time, our addrman may be pretty empty and no one will know who we are.`  less ambiguous:
`If we're starting up for the first time, it would improve our peering chances. Otherwise, our addrman may be pretty empty, and also no one will know who we are.`

But generally, I think this comment:
- makes it much more clear
- actually convinces me that it's an improvement over the existent code.

amitiuttarwar replied:
incorporated these comments into the changes, 

I agree this should be further improved, but I'm hoping this is sufficient for this PR. @naumenkogs are you comfortable with the current state?

jnewbery commented:
Strong strong concept ACK.

I think it'd be good to extract everything in the `while (!interruptNet)` in `ThreadOpenConnections()` into its own function since deeply nested while/for/if blocks obscure control flow and are very often the sources of bugs. Doing that might be scope creep for this PR though.

jnewbery commented on src/net.h:
```diff
@@ -838,37 +838,89 @@ class CNode
+
+    /* Whether we attempt to advertise our own address to the peer when
+     * processing their VERSION message */
+    bool AdvertiseAddressToConn() const {
```
I really dislike these functions being inside `CNode` (same as https://github.com/bitcoin/bitcoin/pull/19316#discussion_r464526492). This is very specific net_processing logic, rather than a general property of the CNode. Having the logic inside net.h rather than at the point it's needed makes it less readable.

amitiuttarwar replied:
yeah, I agree. this is why I initially opted for having `m_conn_type` as a public member var, but feedback on #19316 led me to switching to private. anyway, given the current state, its a bit unfortunate we don't get the additional guarantee of a switch, but I agree with you & incorporated this feedback so it can be more understandable from the call site.

jnewbery commented on src/net.h:
```diff
@@ -838,37 +838,89 @@ class CNode
         assert(false);
     }
 
+    /* These are the default connections that we use to connect with the
```
Format these as doxygen comments (start with `/**`) https://www.doxygen.nl/manual/docblocks.html

amitiuttarwar replied:
oof. so many options and yet I still missed the mark 😅 . moved the comments to the enum, but fixed this problem. thanks!

jnewbery commented on src/net.h:
```diff
@@ -838,37 +838,89 @@ class CNode
         assert(false);
     }
 
+    /* These are the default connections that we use to connect with the
```
These are great comments. Thank you! Should they be merged with the comments above on the enum members?

amitiuttarwar replied:
yeah I felt very indecisive when trying to figure out the best spots for the comments. here what I've gone for is having the enum comments be a more concise description of the nature of the connection, and the function provide more context around design goals, capacity, etc. 

I don't think the split is super clear though, so yeah, maybe it would be more helpful to consolidate it all into the enum member comments. is that what you're advocating for? or just trying to weigh like me? ⚖️

jnewbery replied:
yes, consolidate them to one place. I think the enum is the right place for them.

amitiuttarwar replied:
done!

jnewbery reviewed (commented):
ACK everything except the `AdvertiseAddressToConn()` change. I'd much rather leave the conditional as

`if (!pfrom.IsInboundConn() && !pfrom.IsBlockOnlyConn())`

and add a large comment to above the `connman.MarkAddressGood(pfrom.addr);` to say that this isn't anything to do with sending an addr or getaddr message.

amitiuttarwar committed:
[doc] Follow developer notes, add comment about missing default.

amitiuttarwar commented:
thanks for the feedback! all comments should be addressed. 

@sipa 
I agree with this: 
> make the addr-relay decision depend on the connection type. That should be the most authoritative information about the connection.

But unfortunately I don't think this is true: 
>  but if we accidentally forget to create the necessary data structures, test will fail; 

I updated `RelayAddrsWithConn` to always return true, recompiled & ran `test_runner.py` & unit tests. Nothing failed. My concern would be addressed if it did :) I think with #19315, we should be able to add functional tests that would fail if a discrepancy occurs? 

Currently I think the downside is probably acceptable, but want to highlight since I don't think we have any automatic protection in place. (unless I'm missing something?)

jonatack commented:
Concept ACK on these changes. Late to the party but will try to review soon-ish.

amitiuttarwar commented on src/net_processing.cpp:
```diff
@@ -1982,7 +1982,7 @@ static void ProcessHeadersMessage(CNode& pfrom, CConnman& connman, ChainstateMan
         }
 
-        if (!pfrom.fDisconnect && pfrom.IsOutboundOrBlockRelayConn() && nodestate->pindexBestKnownBlock != nullptr && pfrom.m_tx_relay != nullptr) {
+        if (!pfrom.fDisconnect && pfrom.IsFullOutboundConn() && nodestate->pindexBestKnownBlock != nullptr) {
```
ok, I've looked into this and am happy to discuss with you, but want to clarify we are on the same page that **this PR doesn't change the behavior.**

I agree that the comment initially seems inconsistent with the logic in `ConsiderEviction`, but I'm still trying to wrap my head around all the mechanisms (since the chainwork check is repeated as the first conditional in `ConsiderEviction`, is the protection here essential?). I have many observations & questions, but I don't think this PR is the right place to have this conversation. are you willing to open an issue?

ariard replied:
Yes as far as I understand this code it is neither introduced by this PR or #19316. 

>  (since the chainwork check is repeated as the first conditional in ConsiderEviction, is the protection here essential?).

You mean the check `pindexBestKnownBlock != nullptr` ? But you have a conditional above and it won't exclude block-relay-only with `m_protect=false`. I'm not sure about the exact semantic of `m_protect` but yes I should be able to write a test for this with #19315 to assert bug :)

sdaftuar replied:
Looks like a documentation oversight when I changed around how this code worked while working on #15759. See https://github.com/bitcoin/bitcoin/pull/15759#pullrequestreview-281092623.  I think originally I was going to make block-relay-only peers immune from all eviction, and then changed it to only be immune to outbound peer rotation-based eviction.

ariard replied:
Tracked in https://github.com/bitcoin/bitcoin/issues/19863#issue-692406060, either to document block-relay-only eviction expectations or align protection scope among all our outbound peers.

ariard commented on src/net.h:
```diff
@@ -121,12 +121,54 @@ struct CSerializedNetMsg
+     * Feeler connections are short lived connections used to increase the
+     * number of connectable addresses in our AddrMan. Approximately every
+     * FEELER_INTERVAL, we attempt to connect to a random address from the new
+     * table. If successful, we add it to the tried table.
```
Actually it does also solve bucket conflict between 2 tried tables entries, see : https://github.com/bitcoin/bitcoin/pull/9037

amitiuttarwar replied:
I already updated the comment in `ThreadOpenConnections` with the specifics of the ordering: 

```
        // Determine what type of connection to open. Opening
        // OUTBOUND_FULL_RELAY connections gets the highest priority until we
        // meet our full-relay capacity. Then we open BLOCK_RELAY connection
        // until we hit our block-relay-only peer limit.
        // GetTryNewOutboundPeer() gets set when a stale tip is detected, so we
        // try opening an additional OUTBOUND_FULL_RELAY connection. If none of
        // these conditions are met, check the nNextFeeler timer to decide if
        // we should open a FEELER.
```

and I believe that is the right place for documenting the order. I realized that the block-relay-only enum comment mentioned ordering so I've gone ahead and removed that. I think these comments are most valuable if they focus on the design goals and the high level description of the connection type & we look at the relevant code for more specifics. otherwise we also risk the comments falling out of date. 

---

> Actually it does also solve bucket conflict between 2 tried tables entries, see : #9037

what is this in response to? is there a suggested change?

naumenkogs replied:
> I think these comments are most valuable if they focus on the design goals and the high level description of the connection type & we look at the relevant code for more specifics.

>what is this in response to? is there a suggested change?
Antoine just mentions that there is *another thing* done when considering a feeler. See [here](https://github.com/bitcoin/bitcoin/pull/19724#discussion_r480320438). I think this part, if explained, also belongs there and not here.

amitiuttarwar commented on src/net.cpp:
```diff
@@ -1854,28 +1854,32 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
+        // Determine what type of connection to open. Opening
+        // OUTBOUND_FULL_RELAY connections gets the highest priority until we
+        // meet our full-relay capacity. Then we open BLOCK_RELAY connection
+        // until we hit our block-relay peer limit. GetTryNewOutboundPeer()
```
done

amitiuttarwar commented on src/net.cpp:
```diff
@@ -1854,28 +1854,32 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
+        // gets set when a stale tip is detected, so we try opening an
+        // additional OUTBOUND_FULL_RELAY connection. If none of these
+        // conditions are met, check the nNextFeeler timer to decide if we
+        // should open a feeler.
```
done

amitiuttarwar commented on src/net.cpp:
```diff
@@ -1854,28 +1854,32 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
+        } else if (nTime > nNextFeeler) {
+            nNextFeeler = PoissonNextSend(nTime, FEELER_INTERVAL);
+            conn_type = ConnectionType::FEELER;
+            fFeeler = true;
```
`IsFeelerConn()` is a function on `CNode`, not `ConnectionType`, so I don't think this is possible

sdaftuar replied:
It does seem a little weird to still have fFeeler as a variable after this change, but I don't feel strongly.

naumenkogs replied:
Instead of `if (fFeeler)`, we can now do `if (conn_type == ConnectionType::FEELER)`, and drop the `fFeeler` variable.

naumenkogs replied:
Or, now that I see how many `if (fFeeler)` we have (a lot), I'm not so sure...

amitiuttarwar replied:
yup, I left it in essentially as a local alias. we use it 6 times in the function and it didn't seem helpful to switch those all over to the more verbose `conn_type == ConnectionType::FEELER`

amitiuttarwar commented on src/net.h:
```diff
@@ -121,12 +121,54 @@ struct CSerializedNetMsg
+     * open MAX_BLOCK_RELAY_ONLY_CONNECTIONS to help prevent against partition
+     * attacks. By not relaying transactions or addresses, these connections
+     * are harder to detect by a third party, thus helping obfuscate the
+     * network topology. Addresses are selected from AddrMan.
```
are they always? what happens if your tried table isn't populated yet ?

ariard replied:
Ah sorry, in fact they are drawn from both `new/tried` tables at anytime. Surely being confused by `SelectTriedCollision` in `net.cpp` path!

amitiuttarwar commented on src/net.h:
```diff
@@ -121,12 +121,54 @@ struct CSerializedNetMsg
+    BLOCK_RELAY,
+
+    /**
+     * AddrFetch connections are short lived connections used to solicit
```
feels a bit deep in the weeds for this patch, both have nuances of how disconnection is executed

amitiuttarwar commented on src/net.h:
```diff
@@ -121,12 +121,54 @@ struct CSerializedNetMsg
+    /**
+     * After meeting our quota for full outbound connections, we attempt to
+     * open MAX_BLOCK_RELAY_ONLY_CONNECTIONS to help prevent against partition
+     * attacks. By not relaying transactions or addresses, these connections
```
hmm, seems like our minds are putting parenthesis in different places for this patch. I find "or" clearer. If you feel strongly I can try to rework the sentence. but otherwise, seems fairly trivial to me

amitiuttarwar commented on src/net.h:
```diff
@@ -121,12 +121,54 @@ struct CSerializedNetMsg
+     * After meeting our quota for full outbound connections, we attempt to
+     * open MAX_BLOCK_RELAY_ONLY_CONNECTIONS to help prevent against partition
+     * attacks. By not relaying transactions or addresses, these connections
+     * are harder to detect by a third party, thus helping obfuscate the
```
if two nodes are connected, a third party is any other entity. I don't really understand your suggestion- if a node is "directly paired" with another peer, how would it be possible to obfuscate that connection to the peer? of course the peer can know what connections they have open?

ariard replied:
I think the detection hardness assumption is function on how you define the third-party. 

I meaned "directed paired" as Alice being connected with Bob and Caroll. Caroll guessing the connection Alice-Bob thanks to some relay leaks (like txn or addr).

amitiuttarwar commented on src/net_processing.cpp:
```diff
@@ -2454,6 +2452,9 @@ void ProcessMessage(
             pfrom.fGetAddr = true;
+
+            // Moves address from New to Tried table in Addrman
+            // (unrelated to advertising our address)
```
I feel strongly that we should not cement opinions into docs in the codebase. since there are divergent opinions on whether or not we should have this logic here, I'd like to avoid documentation that indicates it must be here.

naumenkogs replied:
I don't find this comment particularly helpful... Why it is done here, why for outbounds and non-blocks-only?

amitiuttarwar replied:
resolving this conversation just to close off threads, feel free to continue [over here](https://github.com/bitcoin/bitcoin/pull/19724#discussion_r482610113)

amitiuttarwar commented:
thanks for the review @jnewbery & @ariard ! all review comments are addressed 

rebased & made some small changes to docs

sdaftuar reviewed (approved):
utACK, just a couple nits but nothing blocking.

naumenkogs commented on src/net.h:
```diff
@@ -861,6 +861,12 @@ class CNode
     }
 
+    /* Whether we send addr messages over this connection */
+    bool RelayAddrsWithConn() const
```
nit: not a big fan of this name... I actually think `IsAddrRelayPeer` is much better :)
I already saw some discussion around the distinction between "node" and "peer", and now you start referring to it as a "connection".

naumenkogs replied:
Also, this method should be now used here: https://github.com/bitcoin/bitcoin/blob/4c66cc0a5a99cbad51448e51d6b28e11e5232e33/src/net.cpp#L2772

naumenkogs replied:
I think the new code should be:

```
    if (conn_type_in != ConnectionType::BLOCK_RELAY) {
        m_tx_relay = MakeUnique<TxRelay>();
    }
    if (IsAddrRelayPeer()) {
        m_addr_known = MakeUnique<CRollingBloomFilter>(5000, 0.001);
    }
```

amitiuttarwar replied:
ah nice. I like it! I was also concerned about the duplicated logic making room for future discrepancy. I've updated to include this, thanks! 

RE name: seems like an issue with the whole convention around connection types? The name I've chosen is consistent with all the `m_conn_type` functions.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -2426,9 +2426,7 @@ void PeerLogicValidation::ProcessMessage(CNode& pfrom, const std::string& msg_ty
 
-        if (!pfrom.IsInboundConn() && pfrom.IsAddrRelayPeer())
-        {
-            // Advertise our address
```
nit: Why removing the comment? :)

naumenkogs replied:
Sort of justified [here](https://github.com/bitcoin/bitcoin/pull/19724#discussion_r474387266), but i'm not convinced it's the right approach. See my comment there.

amitiuttarwar replied:
the comment removal was leftover from a previous iteration of this code where I updated the conditional  to call a function `AdvertiseAddressToConn` or something that made the comment feel redundant. 

latest push includes the more in depth commenting proposed in that thread, will continue convo there

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -2426,9 +2426,7 @@ void PeerLogicValidation::ProcessMessage(CNode& pfrom, const std::string& msg_ty
-        if (!pfrom.IsInboundConn() && pfrom.IsAddrRelayPeer())
-        {
-            // Advertise our address
+        if (!pfrom.IsInboundConn() && !pfrom.IsBlockOnlyConn()) {
```
This is technically not a behavior change, but I don't understand the logic behind replacing `IsAddrRelayPeer` with `!IsBlockOnlyConn()`. Why not replacing it with `RelayAddrsWithConn`?

naumenkogs replied:
I see it is justified (sort of?) [here](https://github.com/bitcoin/bitcoin/pull/19724#discussion_r474387266).

sdaftuar replied:
The skipping of this logic for block-relay-only outbound peers was introduced in #15759 -- gating it on whether the address-relay data structures are instantiated is confusing because the very last thing we do is unrelated to whether those data structures exist.

Instead, what this block has in common is that we only want to skip it for outbound block-relay only peers, which is what this patch does.  So I think this is right, and while we could split this block into two parts, one gated on whether the peer has the addr-relay data structures, and another gated on whether the peers is an outbound block-relay only connection, that seems pretty unnecessary to me.  If someone adds another connection type that is outbound and non-block-relay only and only needs part of this, it can be refactored then.

jnewbery commented:
reACK 4c66cc0a5a

It looks like all of the remaining review comments are on style/documentation. Perhaps we should aim to merge this and address those in a follow-up?

naumenkogs commented on src/net.cpp:
```diff
@@ -1856,28 +1856,32 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
             }
         }
 
-        // Feeler Connections
```
Can this probably migrate to the enum?

naumenkogs replied:
Partially done in 630b9a5e1872ac91949b1e30f8cd7f28d1c946ab, but some of the details are lost.

amitiuttarwar replied:
what details are lost? the only thing not included is "Start attempting feeler connections only after node finishes making outbound connections," but the ordering of connections is captured in `ThreadOpenConnections`. this seems like the same point as addressed over in [this thread](https://github.com/bitcoin/bitcoin/pull/19724#discussion_r475145927) where you've agreed with me??

naumenkogs commented on src/net.h:
```diff
@@ -118,12 +118,54 @@ struct CSerializedNetMsg
+
+    /**
+     * These are the default connections that we use to connect with the
+     * network. There is no restriction on the inventory relayed- by default we
```
nit: mix of terminology, we normally don't refer to addr as "inventory". Let's keep it that way :)

amitiuttarwar replied:
oh, didn't realize. fixed

naumenkogs commented:
>It looks like all of the remaining review comments are on style/documentation. Perhaps we should aim to merge this and address those in a follow-up?

Most of the changes in this commit are style/documentation. It's not like it's a big logic-changing PR with a lot of ACKs will have to invalidate :)

But actually, I'd be fine with ACKing this one as long as I get some feedback (code change, or at least a comment) on these two comments of mine I consider substantial: [first](https://github.com/bitcoin/bitcoin/pull/19724#discussion_r481094603), [second](https://github.com/bitcoin/bitcoin/pull/19724#discussion_r480960343).

sdaftuar commented:
I think there may be a silent merge conflict here, as a result of merging #19067 a few days ago (after the CI tests had run on this PR).  That PR introduces one new place where the OUTBOUND enum is used.

(Noticed this because of the failing CI run in #19858, built on top of this one.)

amitiuttarwar committed:
[doc] Explain address handling logic in process messages

Co-authored-by: Suhas Daftuar <sdaftuar@gmail.com>

amitiuttarwar commented on src/net_processing.cpp:
```diff
@@ -843,7 +843,7 @@ void PeerLogicValidation::InitializeNode(CNode *pnode) {
         mapNodeState.emplace_hint(mapNodeState.end(), std::piecewise_construct, std::forward_as_tuple(nodeid), std::forward_as_tuple(addr, std::move(addrName), pnode->IsInboundConn(), pnode->IsManualConn()));
     }
-    if(!pnode->IsInboundConn())
+    if (!pnode->IsInboundConn())
```
done

amitiuttarwar commented:
thank you for the reviews! 

good catch on the silent merge conflict @sdaftuar, I’ve rebased to address. (turns out, there was more than one)

I've addressed all review comments. They were getting pretty web-like, so I've tried my best to resolve and focus on any outstanding conversations. 


>> It looks like all of the remaining review comments are on style/documentation. Perhaps we should aim to merge this and address those in a follow-up?

> Most of the changes in this commit are style/documentation. It's not like it's a big logic-changing PR with a lot of ACKs will have to invalidate :)

I'd like to get this PR merged sooner rather than later for a couple reasons: 
* I’d like to get the renames on master to avoid possibility of more silent merge conflicts
* there are now two PRs that build on this (#19858 & #19860). further improvements can be addressed in parallel rather than holding up the whole sequence. 
* I think its important not to discourage review / lose momentum. @jnewbery has ACKed the tip twice already & I’ve had to invalidate because of needing rebase. with 10 commits deep, it doesn't seem trivial to reACK, so I'd rather use that review energy on smaller isolated changes that can be handled in another PR.

just trying to express that I'm feeling cautious about invalidating ACKs based on the reviews so far, so might opt for starting another branch/PR if more non-essential changes are requested.

amitiuttarwar commented:
there's one failing build, but I'm confused by it. nothing in the logs appears to have failed?? 
https://travis-ci.org/github/bitcoin/bitcoin/jobs/723629947

RandyMcMillan commented:
@amitiuttarwar - This PR is based on parent commits that aren't passing - the fail may have nothing to do with your change. I recommend rebasing your commit on a recent passing commit.

naumenkogs commented:
ACK eb1c5d090f4ef844dabc1e8bf3864c1afe1c791c

laanwj commented:
Code review ACK eb1c5d090f4ef844dabc1e8bf3864c1afe1c791c

ariard commented:
post-merge Code Review ACK eb1c5d0
